mavros_extras/src/plugins/wheel_odometry.cpp

```diff
@@ -0,0 +1,525 @@
+	 * Twist info doesn't depend on initial orientation so is published from the very start.
+	 * @param time		measurement's ROS time stamp
+	 */
+	void publish_odometry(ros::Time time)
+	{
+		// Get initial yaw (from IMU)
+		// Check that IMU was already initialized
+		if (!yaw_initialized && m_uas->get_attitude_imu_enu()) {
+			auto quat = m_uas->get_attitude_orientation_enu();
+			tf2::Quaternion q(quat.x, quat.y, quat.z, quat.w);
+			double roll, pitch, yaw;
+			tf2::Matrix3x3 mat(q);
+			mat.getRPY(roll, pitch, yaw);
+
+			// Rotate current pose by initial yaw
+			float cosy = std::cos(yaw);
+			float siny = std::sin(yaw);
+			float x = rpose(0);
+			float y = rpose(1);
+			rpose(0) = x*cosy - y*siny; // x
+			rpose(1) = x*siny + y*cosy; // y
+			rpose(2) += yaw; // yaw
+
+			ROS_INFO_NAMED("wo", "WO: Initial yaw (deg) = %f\n", yaw/M_PI*180.0);
+			yaw_initialized = true;
+		}
+
+		// Orientation (only if we have initial yaw)
+		geometry_msgs::Quaternion quat;
+		if (yaw_initialized) {
+			tf2::Quaternion q;
+			q.setRPY(0.0, 0.0, rpose(2));
+			quat.x = q.x();
+			quat.y = q.y();
+			quat.z = q.z();
+			quat.w = q.w();
+		}
+
+		// Twist
+		geometry_msgs::TwistWithCovariance twist_cov;
+		// linear
+		twist_cov.twist.linear.x = rtwist(0);
+		twist_cov.twist.linear.y = rtwist(1);
+		twist_cov.twist.linear.z = 0.0;
+		// angular
+		twist_cov.twist.angular.x = 0.0;
+		twist_cov.twist.angular.y = 0.0;
+		twist_cov.twist.angular.z = rtwist(2);
+		// covariance
+		ftf::EigenMapCovariance6d twist_cov_map(twist_cov.covariance.data());
+		twist_cov_map.setZero();
+		twist_cov_map.block<3, 3>(0, 0).diagonal() << rtwist_cov(0), rtwist_cov(1), -1.0;
+		twist_cov_map.block<3, 3>(3, 3).diagonal() << -1.0, -1.0, rtwist_cov(2);
+
+		// Publish twist
+		if (twist_send) {
+			auto twist_cov_t = boost::make_shared<geometry_msgs::TwistWithCovarianceStamped>();
+			// header
+			twist_cov_t->header.stamp = time;
+			twist_cov_t->header.frame_id = frame_id;
+			// twist
+			twist_cov_t->twist = twist_cov;
+			// publish
+			twist_pub.publish(twist_cov_t);
+		}
+		// Publish odometry (only if we have initial yaw)
+		else if (yaw_initialized) {
+			auto odom = boost::make_shared<nav_msgs::Odometry>();
+			// header
+			odom->header.stamp = time;
+			odom->header.frame_id = frame_id;
+			odom->child_frame_id = child_frame_id;
+			// pose
+			odom->pose.pose.position.x = rpose(0);
+			odom->pose.pose.position.y = rpose(1);
+			odom->pose.pose.position.z = 0.0;
+			odom->pose.pose.orientation = quat;
+			ftf::EigenMapCovariance6d pose_cov_map(odom->pose.covariance.data());
+			pose_cov_map.setZero();
+			pose_cov_map.block<3, 3>(0, 0).diagonal() << rpose_cov(0), rpose_cov(1), -1.0;
+			pose_cov_map.block<3, 3>(3, 3).diagonal() << -1.0, -1.0, rpose_cov(2);
+			// twist
+			odom->twist = twist_cov;
+			// publish
+			odom_pub.publish(odom);
+		}
+
+		// Publish TF (only if we have initial yaw)
+		if (tf_send && yaw_initialized) {
```

Comment:
You can put this line above (inside the if condition) and just use `if (tf_send) { }`

Reply:
Then TF sending is not reached if `twist_send` is `true`.

Reply:
yeah never mind. I was thinking you wanted them mutually exclusive